File: plugins/code-workflows/skills/api-client-generator/assets/typescript/complete-client.ts
/**
 * Complete API client with all features
 * - Type-safe requests/responses with Zod
 * - Retry with exponential backoff
 * - Rate limiting respect
 * - Request/response logging
 * - Authentication (API key, Bearer, OAuth)
 * - Request caching
 * - Interceptors
 */

import ky, { type KyInstance, type BeforeRequestHook, type AfterResponseHook } from "ky";
import { z } from "zod";

// ============================================================================
// Types & Schemas
// ============================================================================

export interface ClientConfig {
	baseUrl: string;
	auth?: AuthConfig;
	timeout?: number;
	retries?: number;
	cache?: CacheConfig;
	logging?: LoggingConfig;
	interceptors?: InterceptorConfig;
}

export interface AuthConfig {
	type: "apiKey" | "bearer" | "oauth";
	apiKey?: string;
	apiKeyHeader?: string;
	token?: string;
	refreshToken?: () => Promise<string>;
}

export interface CacheConfig {
	enabled: boolean;
	ttl: number; // milliseconds
	methods?: ("GET" | "HEAD")[];
}

export interface LoggingConfig {
	enabled: boolean;
	onRequest?: (method: string, url: string, body?: unknown) => void;
	onResponse?: (status: number, url: string, duration: number) => void;
	onError?: (error: Error, url: string) => void;
}

export interface InterceptorConfig {
	beforeRequest?: BeforeRequestHook[];
	afterResponse?: AfterResponseHook[];
}

// ============================================================================
// Errors
// ============================================================================

export const ApiErrorSchema = z.object({
	code: z.string(),
	message: z.string(),
	details: z.unknown().optional(),
});

export type ApiErrorBody = z.infer<typeof ApiErrorSchema>;

export class ApiError extends Error {
	constructor(
		public status: number,
		public code: string,
		message: string,
		public details?: unknown,
		public retryAfter?: number,
	) {
		super(message);
		this.name = "ApiError";
	}

	static fromResponse(status: number, body: unknown): ApiError {
		const parsed = ApiErrorSchema.safeParse(body);
		if (parsed.success) {
			return new ApiError(status, parsed.data.code, parsed.data.message, parsed.data.details);
		}
		return new ApiError(status, "UNKNOWN", "An unknown error occurred", body);
	}

	get isRetryable(): boolean {
		return [408, 429, 500, 502, 503, 504].includes(this.status);
	}

	get isAuthError(): boolean {
		return [401, 403].includes(this.status);
	}
}

export class ValidationError extends Error {
	constructor(
		message: string,
		public zodError: z.ZodError,
		public path: "request" | "response",
	) {
		super(message);
		this.name = "ValidationError";
	}
}

// ============================================================================
// Result Type
// ============================================================================

export type Result<T, E = ApiError | ValidationError> =
	| { ok: true; data: T; headers: Headers }
	| { ok: false; error: E };

// ============================================================================
// Cache Implementation
// ============================================================================

class RequestCache {
	private cache = new Map<string, { data: unknown; expires: number }>();

	constructor(private ttl: number) {}

	get<T>(key: string): T | undefined {
		const entry = this.cache.get(key);
		if (!entry) return undefined;
		if (Date.now() > entry.expires) {
			this.cache.delete(key);
			return undefined;
		}
		return entry.data as T;
	}

	set(key: string, data: unknown): void {
		this.cache.set(key, { data, expires: Date.now() + this.ttl });
	}

	invalidate(pattern?: string): void {
		if (!pattern) {
			this.cache.clear();
			return;
		}
		for (const key of this.cache.keys()) {
			if (key.includes(pattern)) {
				this.cache.delete(key);
			}
		}
	}
}

// ============================================================================
// Client Factory
// ============================================================================

export function createClient(config: ClientConfig) {
	const {
		baseUrl,
		auth,
		timeout = 30000,
		retries = 3,
		cache: cacheConfig,
		logging,
		interceptors,
	} = config;

	// Initialize cache
	const cache = cacheConfig?.enabled ? new RequestCache(cacheConfig.ttl) : null;
	const cacheMethods = cacheConfig?.methods || ["GET"];

	// Build auth headers
	function getAuthHeaders(): Record<string, string> {
		if (!auth) return {};

		switch (auth.type) {
			case "apiKey":
				return { [auth.apiKeyHeader || "X-API-Key"]: auth.apiKey! };
			case "bearer":
				return { Authorization: `Bearer ${auth.token}` };
			case "oauth":
				return { Authorization: `Bearer ${auth.token}` };
			default:
				return {};
		}
	}

	// Create ky instance
	const client: KyInstance = ky.create({
		prefixUrl: baseUrl,
		timeout,
		retry: {
			limit: retries,
			methods: ["get", "put", "delete", "head", "options"],
			statusCodes: [408, 429, 500, 502, 503, 504],
			backoffLimit: 30000,
			delay: (attemptCount) => Math.min(1000 * 2 ** attemptCount, 30000),
		},
		headers: {
			"Content-Type": "application/json",
			...getAuthHeaders(),
		},
		hooks: {
			beforeRequest: [
				// Logging
				...(logging?.enabled
					? [
							((request) => {
								logging.onRequest?.(request.method, request.url, undefined);
							}) as BeforeRequestHook,
						]
					: []),
				// Custom interceptors
				...(interceptors?.beforeRequest || []),
			],
			afterResponse: [
				// Logging
				...(logging?.enabled
					? [
							((request, _options, response) => {
								logging.onResponse?.(response.status, request.url, 0);
								return response;
							}) as AfterResponseHook,
						]
					: []),
				// Handle 401 with token refresh
				...(auth?.type === "oauth" && auth.refreshToken
					? [
							(async (request, options, response) => {
								if (response.status === 401 && auth.refreshToken) {
									const newToken = await auth.refreshToken();
									auth.token = newToken;
									request.headers.set("Authorization", `Bearer ${newToken}`);
									return ky(request, options);
								}
								return response;
							}) as AfterResponseHook,
						]
					: []),
				// Custom interceptors
				...(interceptors?.afterResponse || []),
			],
			beforeRetry: [
				async ({ request, error, retryCount }) => {
					console.log(`[Retry ${retryCount}] ${request.method} ${request.url}`);

					// Respect Retry-After header
					if (error instanceof ky.HTTPError) {
						const retryAfter = error.response.headers.get("Retry-After");
						if (retryAfter) {
							const delay = Number.parseInt(retryAfter, 10) * 1000;
							await new Promise((resolve) => setTimeout(resolve, delay));
						}
					}
				},
			],
		},
	});

	// Request function
	async function request<TResponse, TRequest = unknown>(
		method: "get" | "post" | "put" | "patch" | "delete",
		path: string,
		options: {
			json?: TRequest;
			searchParams?: Record<string, string | number | boolean>;
			requestSchema?: z.ZodType<TRequest>;
			responseSchema?: z.ZodType<TResponse>;
			skipCache?: boolean;
		} = {},
	): Promise<Result<TResponse>> {
		const { json, searchParams, requestSchema, responseSchema, skipCache } = options;

		// Validate request body
		if (json && requestSchema) {
			const result = requestSchema.safeParse(json);
			if (!result.success) {
				return {
					ok: false,
					error: new ValidationError("Request validation failed", result.error, "request"),
				};
			}
		}

		// Check cache for GET requests
		const cacheKey = `${method}:${path}:${JSON.stringify(searchParams || {})}`;
		if (cache && cacheMethods.includes(method.toUpperCase() as "GET" | "HEAD") && !skipCache) {
			const cached = cache.get<TResponse>(cacheKey);
			if (cached) {
				return { ok: true, data: cached, headers: new Headers() };
			}
		}

		try {
			const startTime = Date.now();
			const response = await client[method](path, {
				json,
				searchParams: searchParams as Record<string, string>,
			});

			const data = await response.json<TResponse>();
			const duration = Date.now() - startTime;

			logging?.onResponse?.(response.status, path, duration);

			// Validate response
			if (responseSchema) {
				const result = responseSchema.safeParse(data);
				if (!result.success) {
					return {
						ok: false,
						error: new ValidationError("Response validation failed", result.error, "response"),
					};
				}

				// Cache successful GET responses
				if (cache && cacheMethods.includes(method.toUpperCase() as "GET" | "HEAD")) {
					cache.set(cacheKey, result.data);
				}

				return { ok: true, data: result.data, headers: response.headers };
			}

			return { ok: true, data, headers: response.headers };
		} catch (err) {
			if (err instanceof ky.HTTPError) {
				const body = await err.response.json().catch(() => ({}));
				const apiError = ApiError.fromResponse(err.response.status, body);

				// Check for rate limit
				const retryAfter = err.response.headers.get("Retry-After");
				if (retryAfter) {
					apiError.retryAfter = Number.parseInt(retryAfter, 10);
				}

				logging?.onError?.(apiError, path);
				return { ok: false, error: apiError };
			}

			const error = err instanceof Error ? err : new Error(String(err));
			logging?.onError?.(error, path);
			return { ok: false, error: new ApiError(0, "NETWORK_ERROR", error.message) };
		}
	}

	return {
		// Raw client
		raw: client,

		// Cache management
		cache: {
			invalidate: (pattern?: string) => cache?.invalidate(pattern),
		},

		// HTTP methods
		get: <T>(path: string, options?: Omit<Parameters<typeof request>[2], "json">) =>
			request<T>("get", path, options),

		post: <TRes, TReq = unknown>(
			path: string,
			json: TReq,
			options?: Parameters<typeof request>[2],
		) => request<TRes, TReq>("post", path, { ...options, json }),

		put: <TRes, TReq = unknown>(
			path: string,
			json: TReq,
			options?: Parameters<typeof request>[2],
		) => request<TRes, TReq>("put", path, { ...options, json }),

		patch: <TRes, TReq = unknown>(
			path: string,
			json: TReq,
			options?: Parameters<typeof request>[2],
		) => request<TRes, TReq>("patch", path, { ...options, json }),

		delete: <T>(path: string, options?: Omit<Parameters<typeof request>[2], "json">) =>
			request<T>("delete", path, options),
	};
}

// ============================================================================
// Usage Example
// ============================================================================

/*
// Define schemas
const CreateUserSchema = z.object({
  name: z.string().min(1),
  email: z.string().email(),
});

const UserSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  email: z.string().email(),
  createdAt: z.string().datetime(),
});

type CreateUser = z.infer<typeof CreateUserSchema>;
type User = z.infer<typeof UserSchema>;

// Create client
const api = createClient({
  baseUrl: "https://api.example.com/v1",
  auth: {
    type: "bearer",
    token: process.env.API_TOKEN!,
  },
  cache: {
    enabled: true,
    ttl: 60000, // 1 minute
  },
  logging: {
    enabled: true,
    onRequest: (method, url) => console.log(`→ ${method} ${url}`),
    onResponse: (status, url, duration) => console.log(`← ${status} ${url} (${duration}ms)`),
    onError: (error, url) => console.error(`✕ ${url}:`, error.message),
  },
});

// Create user with validation
const result = await api.post<User, CreateUser>("/users", {
  name: "John Doe",
  email: "john@example.com",
}, {
  requestSchema: CreateUserSchema,
  responseSchema: UserSchema,
});

if (result.ok) {
  console.log("Created user:", result.data.id);
} else if (result.error instanceof ValidationError) {
  console.log("Validation failed:", result.error.zodError.issues);
} else {
  console.log("API error:", result.error.code);
}

// Invalidate cache after mutation
api.cache.invalidate("/users");
*/

File: plugins/code-workflows/skills/api-client-generator/assets/python/complete_client.py
"""
Complete API client with all features
- Type-safe requests/responses with Pydantic
- Retry with exponential backoff (tenacity)
- Rate limiting respect
- Request/response logging
- Authentication (API key, Bearer, OAuth)
- Request caching
"""

from __future__ import annotations

import hashlib
import logging
import time
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from enum import Enum
from functools import wraps
from typing import Any, Callable, Generic, ParamSpec, TypeVar

import httpx
from pydantic import BaseModel, ValidationError
from tenacity import (
    RetryError,
    retry,
    retry_if_exception_type,
    stop_after_attempt,
    wait_exponential,
)

logger = logging.getLogger(__name__)

T = TypeVar("T", bound=BaseModel)
P = ParamSpec("P")
R = TypeVar("R")


# ============================================================================
# Configuration
# ============================================================================


class AuthType(Enum):
    API_KEY = "api_key"
    BEARER = "bearer"
    OAUTH = "oauth"


@dataclass
class AuthConfig:
    type: AuthType
    api_key: str | None = None
    api_key_header: str = "X-API-Key"
    token: str | None = None
    refresh_token: str | None = None
    token_url: str | None = None
    client_id: str | None = None
    client_secret: str | None = None


@dataclass
class CacheConfig:
    enabled: bool = True
    ttl: int = 300  # seconds
    max_size: int = 1000


@dataclass
class RetryConfig:
    enabled: bool = True
    max_attempts: int = 3
    min_wait: float = 1.0
    max_wait: float = 30.0
    retry_statuses: tuple[int, ...] = (408, 429, 500, 502, 503, 504)


@dataclass
class LoggingConfig:
    enabled: bool = True
    log_body: bool = False
    log_headers: bool = False


@dataclass
class ClientConfig:
    base_url: str
    auth: AuthConfig | None = None
    timeout: float = 30.0
    retry: RetryConfig = field(default_factory=RetryConfig)
    cache: CacheConfig = field(default_factory=CacheConfig)
    logging: LoggingConfig = field(default_factory=LoggingConfig)


# ============================================================================
# Errors
# ============================================================================


class ApiError(Exception):
    def __init__(
        self,
        status: int,
        code: str,
        message: str,
        details: Any = None,
        retry_after: int | None = None,
    ) -> None:
        self.status = status
        self.code = code
        self.message = message
        self.details = details
        self.retry_after = retry_after
        super().__init__(f"[{status}] {code}: {message}")

    @classmethod
    def from_response(cls, response: httpx.Response) -> "ApiError":
        try:
            body = response.json()
            return cls(
                status=response.status_code,
                code=body.get("code", body.get("error", "UNKNOWN")),
                message=body.get("message", body.get("error_description", response.reason_phrase)),
                details=body.get("details"),
                retry_after=int(response.headers.get("Retry-After", 0)) or None,
            )
        except Exception:
            return cls(
                status=response.status_code,
                code="UNKNOWN",
                message=response.reason_phrase,
            )


class RetryableError(Exception):
    """Wrapper for retryable errors"""

    def __init__(self, error: ApiError) -> None:
        self.error = error
        super().__init__(str(error))


class ValidationError_(Exception):
    def __init__(self, message: str, errors: list[dict[str, Any]]) -> None:
        self.errors = errors
        super().__init__(message)


# ============================================================================
# Result Type
# ============================================================================


@dataclass
class Ok(Generic[T]):
    data: T
    headers: dict[str, str] = field(default_factory=dict)
    cached: bool = False

    @property
    def ok(self) -> bool:
        return True


@dataclass
class Err:
    error: ApiError | ValidationError_

    @property
    def ok(self) -> bool:
        return False


Result = Ok[T] | Err


# ============================================================================
# Cache
# ============================================================================


class RequestCache:
    def __init__(self, config: CacheConfig) -> None:
        self.config = config
        self._cache: dict[str, tuple[Any, datetime]] = {}

    def _make_key(self, method: str, url: str, params: dict[str, Any] | None) -> str:
        key_str = f"{method}:{url}:{params}"
        return hashlib.md5(key_str.encode()).hexdigest()

    def get(self, method: str, url: str, params: dict[str, Any] | None = None) -> Any | None:
        if not self.config.enabled:
            return None

        key = self._make_key(method, url, params)
        if key not in self._cache:
            return None

        data, expires = self._cache[key]
        if datetime.now() > expires:
            del self._cache[key]
            return None

        return data

    def set(
        self,
        method: str,
        url: str,
        data: Any,
        params: dict[str, Any] | None = None,
    ) -> None:
        if not self.config.enabled:
            return

        # Evict old entries if cache is full
        if len(self._cache) >= self.config.max_size:
            oldest_key = min(self._cache.keys(), key=lambda k: self._cache[k][1])
            del self._cache[oldest_key]

        key = self._make_key(method, url, params)
        expires = datetime.now() + timedelta(seconds=self.config.ttl)
        self._cache[key] = (data, expires)

    def invalidate(self, pattern: str | None = None) -> None:
        if pattern is None:
            self._cache.clear()
        else:
            keys_to_delete = [k for k in self._cache if pattern in k]
            for key in keys_to_delete:
                del self._cache[key]


# ============================================================================
# Client
# ============================================================================


class ApiClient:
    def __init__(self, config: ClientConfig) -> None:
        self.config = config
        self.cache = RequestCache(config.cache)
        self._client: httpx.Client | None = None
        self._async_client: httpx.AsyncClient | None = None
        self._token_expires: datetime | None = None

    def _get_headers(self) -> dict[str, str]:
        headers = {"Content-Type": "application/json"}

        if self.config.auth:
            match self.config.auth.type:
                case AuthType.API_KEY:
                    headers[self.config.auth.api_key_header] = self.config.auth.api_key or ""
                case AuthType.BEARER | AuthType.OAUTH:
                    headers["Authorization"] = f"Bearer {self.config.auth.token}"

        return headers

    def _refresh_oauth_token(self) -> None:
        """Refresh OAuth token if expired"""
        if not self.config.auth or self.config.auth.type != AuthType.OAUTH:
            return

        if self._token_expires and datetime.now() < self._token_expires:
            return

        if not self.config.auth.refresh_token or not self.config.auth.token_url:
            return

        response = httpx.post(
            self.config.auth.token_url,
            data={
                "grant_type": "refresh_token",
                "refresh_token": self.config.auth.refresh_token,
                "client_id": self.config.auth.client_id,
                "client_secret": self.config.auth.client_secret,
            },
        )

        if response.is_success:
            data = response.json()
            self.config.auth.token = data["access_token"]
            self.config.auth.refresh_token = data.get(
                "refresh_token",
                self.config.auth.refresh_token,
            )
            expires_in = data.get("expires_in", 3600)
            self._token_expires = datetime.now() + timedelta(seconds=expires_in - 60)

    @property
    def client(self) -> httpx.Client:
        if self._client is None:
            self._client = httpx.Client(
                base_url=self.config.base_url,
                headers=self._get_headers(),
                timeout=self.config.timeout,
            )
        return self._client

    def _log_request(
        self,
        method: str,
        url: str,
        body: Any = None,
    ) -> None:
        if not self.config.logging.enabled:
            return
        msg = f"→ {method} {url}"
        if self.config.logging.log_body and body:
            msg += f" {body}"
        logger.info(msg)

    def _log_response(
        self,
        status: int,
        url: str,
        duration: float,
    ) -> None:
        if not self.config.logging.enabled:
            return
        logger.info(f"← {status} {url} ({duration:.2f}s)")

    def _make_request(
        self,
        method: str,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        json: dict[str, Any] | BaseModel | None = None,
        model: type[T] | None = None,
        skip_cache: bool = False,
    ) -> Result[T]:
        # Check cache for GET requests
        if method == "GET" and not skip_cache:
            cached = self.cache.get(method, path, params)
            if cached is not None:
                if model:
                    return Ok(data=model.model_validate(cached), cached=True)
                return Ok(data=cached, cached=True)

        # Refresh OAuth token if needed
        self._refresh_oauth_token()

        # Prepare body
        body = json.model_dump() if isinstance(json, BaseModel) else json

        self._log_request(method, path, body)
        start_time = time.time()

        # Make request with retry
        @retry(
            retry=retry_if_exception_type(RetryableError),
            stop=stop_after_attempt(self.config.retry.max_attempts),
            wait=wait_exponential(
                min=self.config.retry.min_wait,
                max=self.config.retry.max_wait,
            ),
            reraise=True,
        )
        def do_request() -> httpx.Response:
            response = self.client.request(method, path, params=params, json=body)

            if response.status_code in self.config.retry.retry_statuses:
                error = ApiError.from_response(response)
                if error.retry_after:
                    time.sleep(error.retry_after)
                raise RetryableError(error)

            return response

        try:
            response = do_request()
        except RetryError as e:
            if isinstance(e.last_attempt.exception(), RetryableError):
                return Err(e.last_attempt.exception().error)
            raise
        except httpx.RequestError as e:
            return Err(ApiError(status=0, code="NETWORK_ERROR", message=str(e)))

        duration = time.time() - start_time
        self._log_response(response.status_code, path, duration)

        if not response.is_success:
            return Err(ApiError.from_response(response))

        data = response.json()

        # Cache GET responses
        if method == "GET":
            self.cache.set(method, path, data, params)

        # Validate response
        if model:
            try:
                validated = model.model_validate(data)
                return Ok(data=validated, headers=dict(response.headers))
            except ValidationError as e:
                return Err(ValidationError_("Validation failed", e.errors()))

        return Ok(data=data, headers=dict(response.headers))

    # -------------------------------------------------------------------------
    # HTTP Methods
    # -------------------------------------------------------------------------

    def get(
        self,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        model: type[T] | None = None,
        skip_cache: bool = False,
    ) -> Result[T]:
        return self._make_request("GET", path, params=params, model=model, skip_cache=skip_cache)

    def post(
        self,
        path: str,
        *,
        json: dict[str, Any] | BaseModel | None = None,
        model: type[T] | None = None,
    ) -> Result[T]:
        result = self._make_request("POST", path, json=json, model=model)
        # Invalidate related cache
        self.cache.invalidate(path.split("/")[1] if "/" in path else None)
        return result

    def put(
        self,
        path: str,
        *,
        json: dict[str, Any] | BaseModel | None = None,
        model: type[T] | None = None,
    ) -> Result[T]:
        result = self._make_request("PUT", path, json=json, model=model)
        self.cache.invalidate(path.split("/")[1] if "/" in path else None)
        return result

    def patch(
        self,
        path: str,
        *,
        json: dict[str, Any] | BaseModel | None = None,
        model: type[T] | None = None,
    ) -> Result[T]:
        result = self._make_request("PATCH", path, json=json, model=model)
        self.cache.invalidate(path.split("/")[1] if "/" in path else None)
        return result

    def delete(
        self,
        path: str,
        *,
        model: type[T] | None = None,
    ) -> Result[T]:
        result = self._make_request("DELETE", path, model=model)
        self.cache.invalidate(path.split("/")[1] if "/" in path else None)
        return result

    # -------------------------------------------------------------------------
    # Context Manager
    # -------------------------------------------------------------------------

    def __enter__(self) -> "ApiClient":
        return self

    def __exit__(self, *args: Any) -> None:
        if self._client:
            self._client.close()


# ============================================================================
# Usage Example
# ============================================================================

"""
import logging
from pydantic import BaseModel

logging.basicConfig(level=logging.INFO)

class User(BaseModel):
    id: str
    name: str
    email: str

class CreateUser(BaseModel):
    name: str
    email: str

# Create client with full config
client = ApiClient(
    ClientConfig(
        base_url="https://api.example.com/v1",
        auth=AuthConfig(
            type=AuthType.OAUTH,
            token="access-token",
            refresh_token="refresh-token",
            token_url="https://auth.example.com/oauth/token",
            client_id="client-id",
            client_secret="client-secret",
        ),
        retry=RetryConfig(
            max_attempts=3,
            min_wait=1.0,
            max_wait=10.0,
        ),
        cache=CacheConfig(
            enabled=True,
            ttl=300,
        ),
        logging=LoggingConfig(
            enabled=True,
            log_body=True,
        ),
    )
)

# Make requests
result = client.get("/users/123", model=User)
if result.ok:
    print(f"User: {result.data.name}")
    print(f"Cached: {result.cached}")
else:
    print(f"Error: {result.error.message}")

# Create user
create_result = client.post(
    "/users",
    json=CreateUser(name="John", email="john@example.com"),
    model=User,
)
"""
